File: QytetetRuby/lib/especulador.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


module ModeloQytetet
  
  class Especulador < Jugador
    attr_writer :fianza
    
    def initialize(jugador, fianza)
      copia_jugador(jugador)
      @fianza = fianza
      @factor_especulador = 2
    end
    
    def pagar_impuestos(cantidad)
      modificar_saldo(-cantidad/2)
    end
    
    def ir_a_carcel(casilla)
      if !pagar(@fianza)
        @casilla_actual = casilla
        @encarcelado = true
      end
    end
    
    def pagar_fianza(cantidad)
      puede_pagar_fianza = false
      
      if @saldo > cantidad
        modificar_saldo(-cantidad)
        puede_pagar_fianza = true
      end
      
      return puede_pagar_fianza
    end
    
    def convertirme(fianza)
      return self
    end
    
    private :pagar_fianza
    
  end
  
end

File: QytetetJava/src/modeloqytetet/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author david
 */
public class Especulador extends Jugador {
//    static int factorEspeculador = 2;
    private int fianza;
    
    protected Especulador(Jugador jugador, int fianza) {
        super(jugador);
        this.fianza = fianza;
        this.factorEspeculador = 2;
    }
    
    @Override
    protected void pagarImpuestos(int cantidad) {
        modificarSaldo(-cantidad/2);
    }
    
    @Override
    protected void irACarcel(Casilla casilla) {
        if(!pagarFianza(fianza)) {
            setCasillaActual(casilla);
            setEncarcelado(true);
        }
    }
    
    @Override
    protected Especulador convertirme(int fianza) {
        return this;
    }
    
    private boolean pagarFianza(int cantidad) {
        boolean puedePagarFianza = false;
        
        if(getSaldo() > cantidad) {
            modificarSaldo(-cantidad);
            puedePagarFianza = true;
        }            
        
        return puedePagarFianza;
    }
    
}
